test/test_common/utility.cc

```diff
@@ -32,12 +32,14 @@ using testing::GTEST_FLAG(random_seed);
 
 namespace Envoy {
 
-static const int32_t SEED = std::chrono::duration_cast<std::chrono::nanoseconds>(
-                                std::chrono::system_clock::now().time_since_epoch())
-                                .count();
+int32_t getSeed() {
+  return std::chrono::duration_cast<std::chrono::nanoseconds>(
```

Comment:
OK, makes sense (https://isocpp.org/wiki/faq/ctors#static-init-order-on-intrinsics). Can you make this a `static` internally in this function, so that if we instantiate another `TestRandomGenerator` object, we get the same seed? Thanks.

Reply:
This is the previous talk about `static`
https://github.com/envoyproxy/envoy/pull/4288#discussion_r214069863

It will only get called once (per test) when TestRandomGenerator is constructed, and that should be fine.

Reply:
I've replied at https://github.com/envoyproxy/envoy/pull/4288#discussion_r214776579, we can discuss in that thread, thanks.

Reply:
#4288 is now a minimal fix to an unrelated bug, which doesn't touch this seed-construction at all (anymore); my suggestion is to work toward merging there; and discuss the static fiasco here.

I am not understanding why TestRandomGenerator wants to get the same seed on each test run within a single process, but is OK getting different seeds when linked into multiple processes. One way that's confusing is that the coverage run would get a single seed for the entire suite of envoy tests, whereas the unit tests ordinarily would get a different seed for each _test.cc. If there's some rationale why that needs to work that way, it would be good to have a comment.

But in any case, if we really do want to share the seed within the binary, we need to do that in a non-fiasco way :)  Our attention was called to the fiasco by the explicit use of that term in the ASAN test that failed when @zyfjeff switched from std::chrono to absl::Time in the static initialization expression, in an earlier version of #4288 . My suspicion is that the absl::Time function had its own evil static construction and the linker chose to run that after this one. I didn't investigate deeply because I prefer to avoid static initialization involving function calls.

I'm also not entirely clear when the static initialization is lazy (which at least makes order deterministic). I might have thought it would be lazy here but the ASAN error suggested it was not.

gtest does provide a mechanism for complex static initialization: SetUpTestCase. Could we init the seed in that (if it really needs to be static)?

Reply:
I think we're in agreement that we should fix the static initialization order fiasco and that this current PR is roughly the right thing, just trying to zero in on some minor detail.

I agree that we don't need to have every `TestRandomGenerator` have the exact same seed (in fact in the interest of intended random behavior, it would be ideal that if we have more than one instance that the do use slightly different seeds). However, what we _do_ need is the ability to be able to replay deterministically the behavior that is provided from logs.

In this current PR, if we don't have the CLI flag set and have multiple generators, we actually have two dynamically created seeds. Later on, if the user wants to recreate behavior, which of these logged seeds do they use? Either way, the rerun won't exhibit the original behavior.

What I'm suggesting is to make an inner variable of `getSeed()` static. C++ is very clear that this will then only get initialized on first use and solve the static initialization order fiasco. We can chat IRL; I don't think this is controversial.

On a related note, why do we even need this "initialize seed from time" ability? Doesn't gtest always set `random_seed`? Is this for some situation that we are outside of gtest??

Reply:
If you say that from the point of view that you can replay and reproduce the problem, I recognize that SEED should be static.

Reply:
lgtm -- I forgot that SEED was previously a statically initialized variable. So making it a lazy-initialized static in a helper function will solve the fiasco. 

Can you just add a comment that this wants to be a static so there's just one seed specified for replay?

Reply:
Ok